tracer/test/Datadog.Trace.TestHelpers.AutoInstrumentation/CrashReport.cs

```diff
@@ -0,0 +1,153 @@
+    private readonly JObject _json;
+    private readonly ITestOutputHelper _output;
+    private HttpClient _httpClient;
+
+    public CrashReport(string pathToCrashReport, ITestOutputHelper output)
+    {
+        _output = output;
+        _pdbSymbolsFolder = Path.Combine(Path.GetTempPath(), "symbols");
+        _json = JObject.Parse(File.ReadAllText(pathToCrashReport));
+    }
+
+    public void Dispose()
+    {
+        _httpClient?.Dispose();
+    }
+
+    public async Task<IReadOnlyList<string>> ResolveCrashStackTrace()
+    {
+        var result = new List<string>();
+        var stackTrace = (JArray)_json["error"]["stack"]["frames"];
+
+        foreach (var frame in stackTrace)
+        {
+            result.Add(await ResolveSymbol(frame));
+        }
+
+        return result;
+    }
+
+    private async Task<string> ResolveSymbol(JToken frame)
+    {
+        var name = frame["function"].ToString();
+
+        if (frame["build_id_type"]?.ToString() == "PDB")
+        {
+            var buildId = frame["build_id"].ToString();
+            var pdbName = $"{Path.GetFileNameWithoutExtension(name)}.pdb";
+
+            var moduleAddress = ulong.Parse(frame["module_base_address"].ToString().Replace("0x", string.Empty), NumberStyles.HexNumber);
+            var address = ulong.Parse(frame["ip"].ToString().Replace("0x", string.Empty), NumberStyles.HexNumber);
+
+            var resolvedName = await ResolvePdbSymbol(pdbName, buildId, moduleAddress, address);
+
+            if (resolvedName != null)
+            {
+                var modulePath = name.Split('!')[0];
+                var module = Path.GetFileNameWithoutExtension(modulePath);
+
+                name = $"{module}!{resolvedName}";
+            }
+        }
+
+        return name;
+    }
+
+    private async Task<string> ResolvePdbSymbol(string pdbName, string pdbHash, ulong moduleAddress, ulong address)
+    {
+        var pdbFolder = Path.Combine(_pdbSymbolsFolder, pdbName, pdbHash);
+        var pdbPath = Path.Combine(pdbFolder, pdbName);
+
+        if (!_pdbSymbols.TryGetValue(pdbPath, out var symbols))
+        {
+            var fileExists = File.Exists(pdbPath);
+
+            if (!fileExists)
+            {
+                fileExists = await DownloadSymbol(pdbName, pdbHash, pdbFolder);
+            }
+
+            if (fileExists)
+            {
+                using var reader = new SharpPdb.Native.PdbFileReader(pdbPath);
+
+                symbols = reader.Functions
+                                .OrderBy(f => f.RelativeVirtualAddress)
+                                .Select(f => (f.RelativeVirtualAddress, f.CodeSize, f.GetUndecoratedName()))
+                                .ToList();
+            }
+            else
+            {
+                symbols = new();
+            }
+
+            _pdbSymbols.Add(pdbPath, symbols);
+        }
+
+        string name = null;
+
+        foreach (var symbol in symbols)
```

Comment:
Is there any reason to use a `foreach` instead of a `for` when we conditianally break the loop at L112?
```suggestion
        for (in i = 0; i < symbols.Count && symbol.Offset <= address - moduleAddress ; i++ )
```

Reply:
> Is there any reason to use a foreach instead of a for when we conditianally break the loop at L112?

A very good reason is how ugly the for loop is 😂 

> It's also better for performance

Claims like that need data 😉 But also, this is only used in testing code where any tiny performance difference here is going to be massively overshadowed by the fact we just crashed 😂

Reply:
Oh yeah that's a good point 😄

> A very good reason is how ugly the for loop is 😂
Even for my first statement actually, it's better to have a `foreach` for readability sakes in testing code. Please ignore me 😅